sw/host/opentitanlib/src/rom_patch/patch.rs

```diff
@@ -0,0 +1,745 @@
+        let base_addr = u32::from_str(&routine.base_addr)?;
+        if base_addr < patch_base_addr {
+            bail!(
+                "Invalid subroutine base addr {:#x} < {:#x} ",
+                base_addr,
+                patch_base_addr
+            );
+        }
+
+        let mut code = vec![];
+        for insn in &routine.code {
+            let insn_hex = u32::from_str(insn)?;
+            code.push(insn_hex);
+        }
+
+        let offset = PATCH_CODE_SECTION_OFFSET + (base_addr - patch_base_addr) as usize;
+
+        Ok(Self {
+            base_addr,
+            offset,
+            code,
+        })
+    }
+
+    fn len(&self) -> usize {
+        self.code.len() * 4
+    }
+
+    fn offset(&self) -> usize {
+        self.offset
+    }
+
+    fn range(&self) -> Range<u32> {
+        Range {
+            start: self.base_addr,
+            end: self.base_addr + (self.code.len() as u32 * 4),
+        }
+    }
+}
+
+impl fmt::Display for RomPatchSubRoutine {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let range = self.range();
+        write!(
+            f,
+            "\tBase address {:#04x} Range [{:#04x}:{:#04x}]",
+            self.base_addr, range.start, range.end
+        )
+    }
+}
+
+impl Read for RomPatchSubRoutine {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if buf.len() < self.len() {
+            return Err(std::io::Error::from(std::io::ErrorKind::InvalidInput));
+        }
+
+        let mut buf_offset = 0;
+
+        for insn in &self.code {
+            let insn_bytes = insn.to_le_bytes();
+            let raw_insn_bytes = &insn_bytes as *const _ as *const u8;
+            unsafe {
+                ptr::copy_nonoverlapping(raw_insn_bytes, buf[buf_offset..].as_mut_ptr(), 4);
+            }
+            buf_offset += 4;
+        }
+
+        Ok(self.len())
+    }
+}
+
+struct RomPatch<'a> {
+    header: RomPatchHeader,
+    size: usize,
+    code_offset: usize,
+    signature_offset: usize,
+    base_addr: u32,
+    match_table: [RomPatchMatchDescriptor; N_MATCH_DESCRIPTORS],
+    patch_code: Vec<u32>,
+    sub_routines: Vec<RomPatchSubRoutine>,
+    signature_key: &'a RsaPrivateKey,
+}
+
+impl<'a> RomPatch<'a> {
+    fn from_hjson(patch: &RomPatchHjson, signature_key: &'a RsaPrivateKey) -> Result<Self> {
+        let meta = MultiBitBool::Bool4True as u8; // Lock Valid = 1, Program Start = 0
+        let base_addr = u32::from_str(&patch.base_addr)?;
+        let mut remap_addr = base_addr;
+        let mut match_table: [RomPatchMatchDescriptor; N_MATCH_DESCRIPTORS] = Default::default();
+        let mut patch_code = vec![];
+        let mut sub_routines = vec![];
+
+        for (index, region) in patch.regions.iter().enumerate() {
+            let match_pair = RomPatchMatchDescriptor::from_hjson(remap_addr, region)?;
+            match_table[index] = match_pair;
+            remap_addr += (region.patch_code.len() as u32) * 4;
```

Comment:
replace 4 by `std::size_of<u32>()`? Apply throughout.

Reply:
Yes. I have done that throughout all the code.